### PR TITLE
add podReplacementPolicy and terminating field to job api

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add API changes for https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated.
#### Which issue(s) this PR fixes:

Partial Resolve https://github.com/kubernetes/enhancements/issues/3939 

#### Does this PR introduce a user-facing change?

```release-note
add podReplacementPolicy and terminating field to job api
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated.
```

### CODE REVIEW

**Comment by kannon92** [comment]
/assign @deads2k @alculquicondor @mimowo 

To make the API review limited to scope I moved the api changes in https://github.com/kubernetes/kubernetes/pull/117015 to here.

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:147]:
```suggestion
	// when they are terminating (has a metadata.deletionTimestamp) or failed.
```

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:359]:
```suggestion
	// - Failed means to wait until a previously created Pod is fully terminated (has phase Failed or Succeeded)
```

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:362]:
```suggestion
	// When using podFailurePolicy, a null value for podReplacementPolicy is interpreted as Failed and Failed is the
```

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:364]:
```suggestion
	// When not using podFailurePolicy, a null value for podReplacementPolicy is interpreted as TerminatingOrFailed.
```

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:367]:
```suggestion
	// This is an alpha field. Enable JobPodReplacementPolicy to be able to use this field.
```

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:365]:
```suggestion
	// TerminatingOrFailed and Failed are allowed values when podFailurePolicy is not in use.
```

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:407]:
```suggestion
```

**Review by alculquicondor — COMMENTED** [review]
validation.go and strategy.go should be part of this PR, as it is part of the api-review.

**Comment by kannon92** [comment]
> validation.go and strategy.go should be part of this PR, as it is part of the api-review.

Ah!  I pushed up those changes.

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:408]: still pending

File: pkg/apis/batch/validation/validation_test.go
Comment (alculquicondor) [inline pkg/apis/batch/validation/validation_test.go:127]: leave this one nil (so we test that nil passes validation)

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/apis/batch/validation/validation_test.go
Comment (kannon92) [reply pkg/apis/batch/validation/validation_test.go:127]: Sorry, I find this a bit confusing..  

Do you mean leave it as:

```
PodReplacementPolicy: nil
```
or

Just drop the field entirely?

Comment (alculquicondor) [reply pkg/apis/batch/validation/validation_test.go]: they are equivalent and I prefer just dropping the field.

File: pkg/apis/batch/types.go
Comment (kannon92) [reply pkg/apis/batch/types.go]: You got me as I was waiting for the API to generate.  Updated now.

File: pkg/apis/batch/validation/validation_test.go
Comment (kannon92) [reply pkg/apis/batch/validation/validation_test.go]: Ah I see.  Pushed up the last of my changes.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (deads2k) [inline staging/src/k8s.io/api/batch/v1/types.go:364]: While I can appreciate that defaulting based on another field can produce unexpected outcomes when podFailurePolicy changes, we have defaulting to avoid client interpretation like this comment describes.  I think the field needs to be conditionally defaulted to take the burden off of clients and a transformation from not specifying podFailurePolicy to specifying podFailurePolicy will also need to acknowledge/set the podReplacementPolicy to Failed.

**Review by deads2k — CHANGES_REQUESTED** [review]
API, valididation, and strategy look good, but defaulting needs updating to stop every client from having to separately interpret the meaning a missing podReplacementPolicy.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (kannon92) [reply staging/src/k8s.io/api/batch/v1/types.go:364]: That sounds good.  I'll push up a change adding defaulting logic.

Comment (kannon92) [reply staging/src/k8s.io/api/batch/v1/types.go]: Updated. PTAL when you can.

File: pkg/apis/batch/v1/defaults.go
Comment (mimowo) [inline pkg/apis/batch/v1/defaults.go:68]: non-blocking nit: I would suggest to use `podReplacementPolicyPtr` function

File: pkg/registry/batch/job/strategy_test.go
Comment (mimowo) [inline pkg/registry/batch/job/strategy_test.go:47]: non-blocking nit: Can we also here use `podReplacementPolicyPtr` function?

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (mimowo) [inline staging/src/k8s.io/api/batch/v1/types.go:413]:
```suggestion
	// and have deletionTimestamp).
```

Comment (kannon92) [reply staging/src/k8s.io/api/batch/v1/types.go]: updated.

File: pkg/registry/batch/job/strategy_test.go
Comment (kannon92) [reply pkg/registry/batch/job/strategy_test.go]: added.

File: pkg/apis/batch/v1/defaults.go
Comment (kannon92) [reply pkg/apis/batch/v1/defaults.go]: done.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go]: I agree that it would be complicated to interpret the nil value given the dependency on another field, which is why generally we prefer not to do it.

However, given that defaulting happens in every read from etcd, I'm afraid that this would cause a write for every Job object in the store during a control plane upgrade.

On the other hand, I'm not sure what happens when we clear the field in the strategy.

Maybe we can compromise on something: interpret nil as Failed, but set the default if podFailurePolicy is in use (this should at least be a smaller subset of jobs in most clusters).

Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go]: > While I can appreciate that defaulting based on another field can produce unexpected outcomes when podFailurePolicy changes

are the outcomes merely weird, or do they actively break things? e.g. if I apply one job with one podFailurePolicy, this gets defaulted, then I try to reapply the job with a different podFailurePolicy or with no podFailurePolicy, will the defaulted podReplacementPolicy fail validation on update?

Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go]: > I'm afraid that this would cause a write for every Job object in the store during a control plane upgrade.

Do you mean reading would trigger an etcd write, or something else? Newly defaulted fields wouldn't hit etcd until the next actual write of the object (update/patch/etc).

More concerning than triggering writes is returning/persisting data for feature-disabled fields, e.g. alpha fields.

Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go]: FWIIW, podFailurePolicy is immutable, so effectively podReplacementPolicy is immutable if podFailurePolicy is set.

Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go]: > Do you mean reading would trigger an etcd write, or something else? Newly defaulted fields wouldn't hit etcd until the next actual write of the object (update/patch/etc).

Yes, that's what I meant. I though we discussed something like this could happen, but if I'm mistaken, then great, this shouldn't be a problem.

> More concerning than triggering writes is returning/persisting data for feature-disabled fields, e.g. alpha fields.

Even if we clear them in the strategy?

Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go]: > > More concerning than triggering writes is returning/persisting data for feature-disabled fields, e.g. alpha fields.
> 
> Even if we clear them in the strategy?

strategy isn't invoked on read, so if you default it, it'll get returned populated to the user, which we don't want if the field is alpha+feature-disabled

Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go]: What if we set the "default" via mutation in the strategy, based on the feature gate?

Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go]: that could work, or if you do want default-on-read behavior, we do have existing defaults that are conditional on feature enablement

Comment (deads2k) [reply staging/src/k8s.io/api/batch/v1/types.go]: The failing flow looks like
1. create job without a podFailurePolicy, the podReplacementPolicy is defaulted to TerminatingOrFailed to be compatible with existing behavior (every job has a replacement policy)
2. add a podFailurePolicy, but without a podReplacementPolicy.  The previous behavior (TerminatingOrFailed) cannot be honored because it produces inconsistent results (pods must be Failed in order to know if a new one is needed).

To add the podFailurePolicy, you do have a acknowledge (add) the podReplacementPolicy because it's substantively changing the meaning of your job when you add the podFailurePolicy.  If we take the addition of the podFailurePolicy to implicitly mean "silently change the podReplacementPolicy" (not clear to me that we want to), we might have enough information in field metadata to do that in a strategy (do defaults get assigned to fieldmanagers?).

Letting each client take a guess at what nil means seems likely to result in inconsistent outcomes for various clients.

Comment (kannon92) [reply staging/src/k8s.io/api/batch/v1/types.go]: > that could work, or if you do want default-on-read behavior, we do have existing defaults that are conditional on feature enablement

Just so I understand, this would be sufficent?

```
	if utilfeature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
		if obj.Spec.PodReplacementPolicy == nil {
			if obj.Spec.PodFailurePolicy != nil {
				obj.Spec.PodReplacementPolicy = podReplacementPolicyPtr(batchv1.Failed)
			} else {
				obj.Spec.PodReplacementPolicy = podReplacementPolicyPtr(batchv1.TerminatingOrFailed)
			}
		}
	}
```

I see that we have some examples of this in StatefulSet.

Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go]: > FWIIW, podFailurePolicy is immutable, so effectively podReplacementPolicy is immutable if podFailurePolicy is set.

is it also immutable inside cronjob? does the cross-field defaulting also happen inside of a cronjob spec? we probably wouldn't want that

Comment (deads2k) [reply staging/src/k8s.io/api/batch/v1/types.go]: > is it also immutable inside cronjob? does the cross-field defaulting also happen inside of a cronjob spec? we probably wouldn't want that

I haven't checked the code to see how it was implemented. I wasn't expecting it to impact cronjobs since the job that gets created in the end will get defaulted.

Comment (deads2k) [reply staging/src/k8s.io/api/batch/v1/types.go]: Coded to impact only jobs and the defaulting test for cronjob shows no defaulting: https://github.com/kubernetes/kubernetes/blob/00c7c4d4ee5cc7bea6bcc3768463badd8fa2afbe/pkg/apis/batch/v1/defaults_test.go#L425

Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go]: yup, CronJob defaulting it's a separate function. Also CronJob update validation is a separate function.

We are good to use the feature gate then in job defaulting, then :)

**Comment by deads2k** [comment]
/lgtm
/approve
/hold

I think the defauting thread is resolved, but holding fora bit in case @liggitt didn't considered it resolved.

**Comment by kannon92** [comment]
> /lgtm /approve /hold
> 
> I think the defauting thread is resolved, but holding fora bit in case @liggitt didn't considered it resolved.

I think @alculquicondor or I would hold it anyway while we are still reviewing the implementation PR.

**Comment by liggitt** [comment]
The combination of feature-gated defaulting and limiting the cross-field defaulting to Job where the field is immutable sounds fine. No hold needed for me

**Comment by kannon92** [comment]
> The combination of feature-gated defaulting and limiting the cross-field defaulting to Job where the field is immutable sounds fine. No hold needed for me

@deads2k and @liggitt thank you both for your review!

**Comment by deads2k** [comment]
> I think @alculquicondor or I would hold it anyway while we are still reviewing the implementation PR.

Alright, at your discretion then.

**Comment by alculquicondor** [comment]
please squash

**Comment by kannon92** [comment]
> please squash

done.

File: pkg/apis/batch/types.go
Comment (atiratree) [inline pkg/apis/batch/types.go:149]: these definitions should be the same as in PodReplacementPolicy field, no?

File: pkg/apis/batch/types.go
Comment (atiratree) [inline pkg/apis/batch/types.go:359]: should this mention terminated/succeeded as well?

the failed has the same semantic as in `Failed` where we mention it, right?

```
	// - Failed means to wait until a previously created Pod is fully terminated (has phase
	//   Failed or Succeeded) before creating a replacement Pod.
```


It might be good to also make a short explanation about the problem with succeeded counting against failed: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated/README.md#the-default-job-controller-behavior

Comment (alculquicondor) [reply pkg/apis/batch/types.go:359]: I think it's too much detail for the API comment. But we should definitely leave a note in the documentation.

File: pkg/apis/batch/types.go
Comment (kannon92) [reply pkg/apis/batch/types.go]: Good call!  Pushed up a fix.

File: pkg/apis/batch/types.go
Comment (kannon92) [reply pkg/apis/batch/types.go]: I'll add that to the my list in the doc.

**Comment by alculquicondor** [comment]
you will need to rebase after #118009

**Comment by kannon92** [comment]
> you will need to rebase after #118009

I expect to be rebasing a lot. :(. 

Thanks for the notice though!

**Comment by Atharva-Shinde** [comment]
[exception request received](https://groups.google.com/a/kubernetes.io/g/release-team/c/aP91c__LHLM/m/ui6kP4B9AwAJ).
Holding for it's approval.
/hold

**Comment by kannon92** [comment]
Exception was approved and release team gave me the goahead to cancel the hold.  https://groups.google.com/g/kubernetes-sig-release/c/8m9_MloKhoI

/hold cancel

**Pull request merged**
